### PR TITLE
Add `cargo tree` command.

### LINKED ISSUE #7286: Add a way to pretty-print the dependency tree
Currently I need to:

```
cargo install cargo-tree
cargo tree
``` 

to pretty-print the dependency tree - which on first install takes a while, and because of `cargo-tree` dependency on cargo, needs to often be recompiled when Rust is upgraded.

I shouldn't need to install a cargo subcomman to print the whole dependency tree of a crate in a human readable form. Feels like a way too common task that deserves upstream support.

### PR DESCRIPTION
**Pull request by ehuss**
This migrates [cargo-tree](https://github.com/sfackler/cargo-tree/) into Cargo as a built-in command. This is based on a recent master (https://github.com/sfackler/cargo-tree/tree/4108d216ec072d2e7c9befb4de32175f97a9dbc4), and should be mostly similar in functionality. There are a variety changes:

* `--all-targets` renamed to `--no-filter-targets` to avoid confusion with the `--all-targets` flag used in other Cargo commands with a different meaning.
* `--all`/`-a` renamed to `--no-dedupe` to avoid confusion with the `-all` flag which means "all workspace crates" in other Cargo commands.
* `--duplicate` renamed to `--duplicates` (with alias), just a personal preference.
* Added support for multiple roots (workspace support).
* Added the `--graph-features` flag for including features in the graph (to "explain" why a feature is enabled).
* Added `{f}` to format string to show features.
* Handles new feature resolver.
* Handles cyclical dev dependencies.
* Added a test suite.
* Dropped the dependency on petgraph, in favor of a simpler custom graph.

Closes #7286.

### CODE REVIEW

**Comment by ehuss** [comment]
cc @sfackler

This is a proposal to merge `cargo-tree` as a built-in command.

**Motivation**
Currently, Cargo does not have a built-in method to display or visualize the dependency graph. The information is available in JSON (or Cargo.lock), but not in a human-readable form. The `cargo-tree` command provides a way to see the dependencies. It also has a variety of options for performing useful actions, such as displaying duplicate dependencies.

`cargo-tree` has been available as a Cargo plugin for over 4 years. Making it a built-in command reduces the user's burden for installing and updating a separate plugin. It also helps ensure that the command stays compatible with all changes in Cargo.

**Drawbacks**
This adds to the maintenance burden on the Cargo team.

**Alternatives**
We could leave `cargo-tree` as an external plugin. There are a variety of different plugins available for visualizing and querying dependencies, and we could leave it up to the user to pick the one they want.

There are a variety of other ways to visualize a dependency graph. Using graphviz is a common alternate solution. It is my opinion that the low-tech solution of formatting the tree in the terminal offers a simpler solution that is more broadly accessible. Additionally, large graphs can rapidly become unreadable when naively rendered in graphviz, requiring additional effort from the user to prune or organize it. Alternate output formats can be added at a later date.

**Prior Art**
Several package managers offer similar functionality.

* (Javascript) `npm ls` — Nearly equivalent to `cargo-tree`.
* (Javascript) `yarn list` — Nearly equivalent to `cargo-tree`.
* (Go) `dep status -dot` option will emit graphviz.
* (Go) `go list` can emit dependencies/imports, not in tree form.
* (.NET) — Visual Studio has built-in methods to graphically browse the dependency tree. `nuget` and `dotnet` have ways to list dependencies, though not in a hierarchical fashion that I can see.
* (Haskell) `stack dot` will emit a graphviz graph of dependencies.
* (Julia) The `pkg` REPL can interactively inspect the graph, but doesn't have a tree view I can see.
* (Ruby) `bundle viz` will emit a graphviz PNG.
* (Dart) `pub deps` emits an ASCII tree.

Most dependency managers have additional third-party tools for displaying and graphing dependencies.

**Future possibilities**
* Any of the enhancements listed at https://github.com/sfackler/cargo-tree/issues/
* Make package downloading lazy, so that extra packages are only downloaded if needed. Currently it downloads all packages (even if not displayed or needed), just for simplicity. (This is no different from the `cargo-tree` plugin, which uses `cargo metadata` which forces downloading everything.)

**Questions**
* Any hard objection to the backwards-incompatible changes to the flags mentioned above (`--all` and `--all-targets`)? I think maintaining consistency with the rest of Cargo justifies the temporary disruption when users upgrade.
* Instead of adding `--no-filter-targets`, maybe another option would be to make a special target like `--target=all` to mean all targets?
* Are there any other changes we should make before merging?
  * A more general way to select dependency kinds? Currently it has `--no-dev-dependencies`, but there are requests like <https://github.com/sfackler/cargo-tree/issues/51> to also skip build dependencies. I was thinking of a flag like `--dep-kinds=…` where you can list the dep kinds you want (maybe with negative options like `--dep-kinds=-dev` or `--dep-kinds=no-dev`). This could be useful for other commands (like `cargo doc`?).
  * A more general way to select the prefix format? Currently it has `--prefix-depth` and `--no-indent`. Maybe a single flag like `--prefix=depth` and `--prefix=no-indent`?
    * (BTW, I don't really know the use case for changing the prefix).
  * Any of the other issues listed at <https://github.com/sfackler/cargo-tree/issues>?
  * The format string is a little strange. It doesn't handle optional values very well. It also has some parsing bugs. I'm not sure if we should make changes or not?
* Bikeshed any flag names. `--graph-features`?

**Comment by sfackler** [comment]
As one other bit of prior art, I based the original implementation off of Gradle's dependencies task: https://docs.gradle.org/current/userguide/viewing_debugging_dependencies.html

It seems like a fine time to just make breaking changes to the CLI interface - people will have to opt-in anyway by deleting the old cargo-tree.

WRT to dependency kinds, it's kind of an interesting question. I don't think you really ever want to say "show me all dev dependencies recursively" - you'd want to see *your* dev dependencies, their regular (and maybe build dependencies) etc. It might make more sense to structure it around "what gets built if I run this kind of operation" than specific dependency kinds.

**Comment by metajack** [comment]
I often want to query the first party deps or the third party ones only. `--kind workspace` and `--kind thirdparty` in cargo guppy. I've also added `--kind directthirdparty` now to get only the direct dependencies (a depth isn't sufficient as there are many edges from first party into first party deps). I've also found it useful to omit parts of the tree (for instance you might want to ignore edges into rustc-workspace-hack).

File: src/cargo/ops/tree/format/mod.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+        for raw in Parser::new(format) {
+            let chunk = match raw {
+                RawChunk::Text(text) => Chunk::Raw(text.to_owned()),
+                RawChunk::Argument("p") => Chunk::Package,
+                RawChunk::Argument("l") => Chunk::License,
+                RawChunk::Argument("r") => Chunk::Repository,
+                RawChunk::Argument("f") => Chunk::Features,
+                RawChunk::Argument(a) => {
+                    return Err(anyhow!("unsupported pattern `{}`", a));
```
Comment (alexcrichton) [inline src/cargo/ops/tree/format/mod.rs:30]: This can likely be simply `bail!(...)`

File: src/cargo/ops/tree/format/parse.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
+use std::iter;
+use std::str;
+
+pub enum RawChunk<'a> {
+    Text(&'a str),
+    Argument(&'a str),
+    Error(&'static str),
+}
+
+pub struct Parser<'a> {
```
Comment (alexcrichton) [inline src/cargo/ops/tree/format/parse.rs:10]: Could this have a brief comment of the mini-language it's parsing?

File: src/cargo/ops/tree/format/mod.rs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        match node {
+            Node::Package {
+                package_id,
+                features,
+                ..
+            } => {
+                let package = self.graph.package_for_id(*package_id);
+                for chunk in &self.pattern.0 {
+                    match *chunk {
+                        Chunk::Raw(ref s) => fmt.write_str(s)?,
```
Comment (alexcrichton) [inline src/cargo/ops/tree/format/mod.rs:67]: Nowadays this can probably be modernized with modern match syntax that avoids `ref` and such

File: src/cargo/ops/tree/format/mod.rs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            Node::Feature { name, node_index } => {
+                let for_node = self.graph.node(*node_index);
+                match for_node {
+                    Node::Package { package_id, .. } => {
+                        write!(fmt, "{} feature \"{}\"", package_id.name(), name)?;
+                        if self.graph.is_cli_feature(self.node_index) {
+                            write!(fmt, " (command-line)")?;
+                        }
+                    }
+                    _ => panic!("unexpected feature node {:?}", for_node),
```
Comment (alexcrichton) [inline src/cargo/ops/tree/format/mod.rs:101]: Could this have a comment pointing to where it's otherwise validated this doesn't come up?

File: src/cargo/ops/tree/graph.rs [hunk 5]
```diff
@@ -0,0 +1,562 @@
...
+        if !indexes.contains(&index) {
+            indexes.push(index)
+        }
+    }
+}
+
+/// A graph of dependencies.
+pub struct Graph<'a> {
+    nodes: Vec<Node>,
+    edges: Vec<Edges>,
```
Comment (alexcrichton) [inline src/cargo/ops/tree/graph.rs:55]: Having only read this far so far, I'm actually not entirely certain how this representation is going to work out. Could this have a comment for how these two lists correspond to each other?

File: src/cargo/ops/tree/graph.rs [hunk 6]
```diff
@@ -0,0 +1,562 @@
...
+}
+
+#[derive(Debug, Copy, Hash, Eq, Clone, PartialEq)]
+pub enum Edge {
+    Dep(DepKind),
+    Feature,
+}
+
+#[derive(Clone)]
+struct Edges(HashMap<Edge, Vec<usize>>);
```
Comment (alexcrichton) [inline src/cargo/ops/tree/graph.rs:37]: In addition to the comments below on `edges`, I'm a little lost here (again I haven't read the whole module, so this is a comment from just reading the data structures) about why there's a `Vec<HashMap<_, Vec>>` here, so mind adding a comment here explaining a bit?

File: src/cargo/ops/tree/graph.rs [hunk 7]
```diff
@@ -0,0 +1,562 @@
...
+        self.edges.push(Edges::new());
+        self.index
+            .insert(self.nodes[from_index].clone(), from_index);
+        from_index
+    }
+
+    /// Returns a list of nodes the given node index points to for the given kind.
+    ///
+    /// Returns None if there are none.
+    pub fn connected_nodes(&self, from: usize, kind: &Edge) -> Option<Vec<usize>> {
```
Comment (alexcrichton) [inline src/cargo/ops/tree/graph.rs:97]: Take this with a grain of salt since I'm still reading, but could this return `Vec<usize>` instead of `Option<Vec<_>>` where empty means "no connections"?

File: src/cargo/ops/tree/graph.rs (hunk 5 above)
Comment (alexcrichton) [reply src/cargo/ops/tree/graph.rs:55]: (reading a bit more, could this be one list of tuples?)

File: src/cargo/ops/tree/graph.rs [hunk 8]
```diff
@@ -0,0 +1,562 @@
...
+            .nodes
+            .iter()
+            .enumerate()
+            .filter(|(_i, node)| match node {
+                Node::Package { package_id, .. } => package_ids.contains(package_id),
+                _ => false,
+            })
+            .map(|(i, node)| (node, i))
+            .collect();
+        result.sort_unstable();
```
Comment (alexcrichton) [inline src/cargo/ops/tree/graph.rs:126]: Could this have a comment explaining why it's sorting? (and why unstable is ok here vs a normal `sort()`?)

File: src/cargo/ops/tree/graph.rs [hunk 9]
```diff
@@ -0,0 +1,562 @@
...
+    name: InternedString,
+    from: Option<usize>,
+    to: usize,
+    kind: Edge,
+) -> usize {
+    let node = Node::Feature {
+        node_index: to,
+        name,
+    };
+    let (node_index, _new) = match graph.index.get(&node) {
```
Comment (alexcrichton) [inline src/cargo/ops/tree/graph.rs:410]: `_new` is unused I think?

File: src/cargo/ops/tree/mod.rs [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+    };
+
+    for (i, root_index) in roots.into_iter().enumerate() {
+        if i != 0 {
+            println!();
+        }
+
+        // The visited deps is used to display a (*) whenever a dep has
+        // already been printed (ignored with --no-dedupe).
+        let mut visited_deps = HashSet::new();
```
Comment (alexcrichton) [inline src/cargo/ops/tree/mod.rs:192]: Perhaps a minor point, but how come this is reset between each root?

File: src/doc/man/cargo-tree.adoc [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
++
+- `{p}` — The package name.
+- `{l}` — The package license.
+- `{r}` — The package repository URL.
+- `{f}` — Comma-separated list of package features that are enabled.
+
+*--no-indent*::
+    Display the dependencies as a list (rather than a tree).
+
+*--prefix-depth*::
```
Comment (alexcrichton) [inline src/doc/man/cargo-tree.adoc:105]: I'd be in favor of consolidating options where possible, like your suggestion here to merge `--prefix-depth` and `--no-indent`

File: src/doc/man/cargo-tree.adoc [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+== OPTIONS
+
+=== Tree Options
+
+*-i*::
+*--invert*::
+    Invert the tree. Typically this flag is used with the `-p` flag to show
+    the dependencies for a specific package.
+
+*--no-dedupe*::
```
Comment (alexcrichton) [inline src/doc/man/cargo-tree.adoc:48]: Over time I've tended to prefer "positive" CLI flags to negative ones. The main reason is more rustc-specific when these flags take values, but I find double-negatives really hard to read (`-C no-foo=no` means "foo=yes" actually). Do we want to consider renaming these flags so it's something like `--deduplicate=no` or `--dev-dependencies=no` perhaps?

File: src/doc/man/cargo-tree.adoc [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+only one instance is built.
+
+*--no-dev-dependencies*::
+    Do not include dev-dependencies.
+
+*--target* _TRIPLE_::
+    Filter dependencies matching the given target-triple.
+    The default is the host platform.
+
+*--no-filter-targets*::
```
Comment (alexcrichton) [inline src/doc/man/cargo-tree.adoc:73]: I personally like your idea of `--target=all` as a special target which says don't filter. It'd be sort of weird to do `--target foo --no-filter-targets` anyway

File: src/doc/man/cargo-tree.adoc [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+    `--invert`). When used with the `-p` flag, only shows duplicates within
+    the subtree of the given package.
++
+It can be beneficial for build times and executable sizes to avoid building
+that same package multiple times. This flag can help identify the offending
+packages. You can then investigate if the package that depends on the
+duplicate with the older version can be updated to the newer version so that
+only one instance is built.
+
+*--no-dev-dependencies*::
```
Comment (alexcrichton) [inline src/doc/man/cargo-tree.adoc:66]: Your recommendation of `--dep-kinds` makes sense to me, and I think that'd lend itself pretty nicely here (and avoid the no-* prefix!)

File: src/doc/man/cargo-tree.adoc [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+
+*--target* _TRIPLE_::
+    Filter dependencies matching the given target-triple.
+    The default is the host platform.
+
+*--no-filter-targets*::
+    Show dependencies for all target platforms. Cannot be specified with
+    `--target`.
+
+*--graph-features*::
```
Comment (alexcrichton) [inline src/doc/man/cargo-tree.adoc:77]: This is so darn useful it may be nice to have a short flag for this as well perhaps? I guess `-f` is already taken, but we could always be semi-random and do like `-x`

**Comment by alexcrichton** [comment]
So a the recommendation of @ehuss I decided to give this a spin locally to see if I could debug an actual issue I had in builds. One thing I've noticed is that in [the wasmtime repo](https://github.com/bytecodealliance/wasmtime) these two commands cause quite a few rebuilds:

```
$ cargo build --manifest-path crates/api/Cargo.toml
$ cargo test --all --exclude lightbeam --exclude wasmtime-c-api --no-run
```

Using another tool of mine I got output that looks like this:

<details>

```
duplicate artifacts found when compiling, this typically means that something was recompiled because a transitive dependency has different features activated than in a previous build:

the following dependencies are duplicated although they have the same features enabled:
the following dependencies have different features:
  regex-syntax 0.6.17 (registry+https://github.com/rust-lang/crates.io-index)
    build1 additionally enabled features {"unicode", "default"} at "/home/alex/code/wasmtime/target/debug/deps/libregex_syntax-37d0dc3aa6c97f66.rlib"
    build0 additionally enabled features {} at "/home/alex/code/wasmtime/target/debug/deps/libregex_syntax-025cc46211ab3255.rlib"
  syn 1.0.17 (registry+https://github.com/rust-lang/crates.io-index)
    build1 additionally enabled features {"extra-traits"} at "/home/alex/code/wasmtime/target/debug/deps/libsyn-5c3d8b7c7d3bf908.rlib"
    build0 additionally enabled features {} at "/home/alex/code/wasmtime/target/debug/deps/libsyn-edabbd70c7a2db6e.rlib"
  object 0.18.0 (registry+https://github.com/rust-lang/crates.io-index)
    build1 additionally enabled features {"indexmap", "write_core", "crc32fast", "write"} at "/home/alex/code/wasmtime/target/debug/deps/libobject-70dfec46d12ad840.rlib"
    build0 additionally enabled features {} at "/home/alex/code/wasmtime/target/debug/deps/libobject-af359f2409925054.rlib"
  cranelift-codegen 0.61.0 (path+file:///home/alex/code/wasmtime/cranelift/codegen)
    build1 additionally enabled features {"all-arch", "riscv", "x86", "arm32", "arm64", "testing_hooks"} at "/home/alex/code/wasmtime/target/debug/deps/libcranelift_codegen-1a33ce6f0cb8fa3b.rlib"
    build0 additionally enabled features {} at "/home/alex/code/wasmtime/target/debug/deps/libcranelift_codegen-82466da147164ef8.rlib"

builds traced are:

	build0 - ["build", "--manifest-path", "crates/api/Cargo.toml"]
	build1 - ["test", "--all", "--exclude", "lightbeam", "--exclude", "wasmtime-c-api", "--no-run"]
```

</details>

So here's some things I did given this output:

### `regex-syntax`

No idea why these features are changing.

First I ran `cargo tree --graph-features -p regex-syntax -i --manifest-path crates/api/Cargo.toml` [giving this](https://gist.github.com/alexcrichton/0d4db9ba34da4369d41affd52e28ab2d). Sure enough the `regex-syntax` crate doesn't have the `"default"` or `"unicode"` features activated.

Next I wasn't actually sure what next to do since there weren't `--all` analogues for `cargo tree`. Maybe we should have similar package selection flags for `cargo tree`? Additionally I wasn't actually sure how I'd handle `cargo test` or not. I suspect that it's from `[dev-dependencies]` though?

From prior knowledge I know the `crates/wiggle` crate depends on `proptest` which enables these features, and sure enough going into that directory and executing `cargo tree --graph-features -p regex-syntax -i` I immediately found `proptest` through `[dev-dependencies]` as the culprit.

So I think my conclusion from this would be that the package selection arguments `--all` and such may want to be added? (basically expanding the set of roots)

### `syn`

Next I wanted to check `syn` and where it got `extra-traits` from. Using `cargo tree` in `crates/api` it sure enough shows no `extra-traits` anywhere.

Like before I didn't know what to do next, but I had prior knowledge that this is due to the `derive_arbitrary` crate's dependency on `synstructure`, so going into `crates/fuzzing` I could immediately see this.

### `object `

Out of idle curiosity I looked into `object`. Again I had prior knowledge of where to look but it was immediately obvious who enabled the extra features in the test builds once I ran `cargo tree` there.

---

Basically I found this really useful for my use case of figuring out "why was this enabled?" It'd be helped to support the same `--all` family of package selection arguments but I suspect that isn't too hard. It was a lot of output and I found myself leaning on `--no-dedupes` to make sense of anything. My personal use case was "given this node in the graph print me a singular path to the root" so it was easiest to `--no-dedupes`, pipe to `less`, search, and follow visually to the end.

Overall I'm still quite excited for this :)

**Comment by sfackler** [comment]
With respect to "why am I rebuilding things" questions, I think they're common enough to potentially have a more tailored tool specifically for that question. Trying to track the answer down by manually diffing two feature activation trees seems awkward when ideally you'd be able to directly ask cargo "what's changing in my transitive dependency feature selection between operation X and operation Y?".

**Comment by alexcrichton** [comment]
I agree yeah my use case here specifically wasn't really targeted for `cargo tree`, but I can at least say that despite having so much information in `cargo tree --graph-features` I was still able to find what I needed pretty quickly.

File: src/cargo/ops/tree/format/parse.rs (hunk 2 above)
Comment (ehuss) [reply src/cargo/ops/tree/format/parse.rs:32]: The parser seems a little more elaborate than what is necessary (as compared to a simple `.replace()` substitution).  Perhaps @sfackler had grander plans for it?

File: src/cargo/ops/tree/graph.rs (hunk 5 above)
Comment (ehuss) [reply src/cargo/ops/tree/graph.rs:68]: Hm, I tried tuples, but the code ends up being littered with `.0` and `.1` tuple indexes all over, and I think it reduced the readability.  I also kinda tried a struct wrapper, but that also seemed to make it more convoluted.  Dunno, do you feel strongly it needs to change?

File: src/cargo/ops/tree/graph.rs (hunk 7 above)
Comment (ehuss) [reply src/cargo/ops/tree/graph.rs]: Yea, originally it was returning a reference and couldn't easily return an empty vec, but I later changed it (when I added sorting).

File: src/cargo/ops/tree/mod.rs (hunk 10 above)
Comment (ehuss) [reply src/cargo/ops/tree/mod.rs]: No good reason.  I changed it so that it de-dupes across roots.  I think that should be OK.

File: src/doc/man/cargo-tree.adoc (hunk 11 above)
Comment (ehuss) [reply src/doc/man/cargo-tree.adoc]: Consolidated into `--prefix`.

File: src/doc/man/cargo-tree.adoc (hunk 12 above)
Comment (ehuss) [reply src/doc/man/cargo-tree.adoc:48]: I agree that avoiding double negatives is good, but in this case I don't think it is important. There isn't a corresponding flag for `dedupe` since that is the default.  I don't feel like this needs to be super flexible, so I don't think it should take a value (and swapping the "no" from the front to the back doesn't substantially change things).

I do agree the flag itself is potentially confusing.  Other alternatives I thought of was `--show-dupes` or `--show-duplicates`, but it's starting to get long and wordy.  I was also concerned about confusion with the `--duplicates` flag, which means something completely different.

One (tangentially related) idea I wanted to toss out is changing `(*)` to `deduped`.  This is what NPM uses, so there might be some familiarity there.  With the explicit wording, it might make more sense?  I mention this because I remember the first time I used `cargo tree` I was confused what the `(*)` signified.

Any other ideas?

File: src/doc/man/cargo-tree.adoc (hunk 13 above)
Comment (ehuss) [reply src/doc/man/cargo-tree.adoc]: Added `--target=all`.

It also made me realize that `--target=all` is broken with `-Zfeatures=itarget`.  Probably not a high priority, but I'll file an issue later.

File: src/doc/man/cargo-tree.adoc (hunk 14 above)
Comment (ehuss) [reply src/doc/man/cargo-tree.adoc]: I added `--dep-kinds` which supports a variety of combinations.  I'm not 100% happy with it (`--dep-kinds=no-dep` is wordy and awkward).  But I wonder if maybe `--dep-kinds=normal` might be a more common usage, and is maybe a little easier to read?

File: src/doc/man/cargo-tree.adoc (hunk 15 above)
Comment (ehuss) [reply src/doc/man/cargo-tree.adoc:84]: Sure.  `-x` is a little strange.  I'm really tempted just to steal `-f`.  How awful would that be?

**Comment by ehuss** [comment]
> there weren't `--all` analogues for `cargo tree`. Maybe we should have similar package selection flags for `cargo tree`?

It does have package selection flags (`-p`, `--workspace`, `--exclude`).  I excluded `--all` since in the original `cargo tree` it meant something different (`--no-dedupe`) and Cargo's `--all` has been deprecated in favor of `--workspace` to avoid confusion.  I added an error message if you pass `--all` that explains the ambiguity.

But perhaps I'm not understanding what you mean here?

> Additionally I wasn't actually sure how I'd handle `cargo test` or not. I suspect that it's from `[dev-dependencies]` though?

I think the answer here is yes, though I'm not sure what you mean.  Features only differ if you have `-Zfeatures=dev_dep`, but I don't think that's what you mean.

**Comment by ehuss** [comment]
There are plenty of good ideas here (like limiting to "workspace only", or being able to prune parts of the tree).  I'd like to land this first before doing too many more major additions.  I'll keep track of them and file issues later.

File: src/cargo/ops/tree/format/parse.rs (hunk 2 above)
Comment (sfackler) [reply src/cargo/ops/tree/format/parse.rs:32]: I think I just stole most of it from the fmt syntax parser in the standard library.

Comment (alexcrichton) [reply src/cargo/ops/tree/format/parse.rs:32]: Oh I think it's fine and probably inevitable, string-replacement can't do things like `{{t}}` generating the text `{t}` without doing `{$replacement_for_t}`, and it's all small and containe enough anyway.

File: src/cargo/ops/tree/graph.rs (hunk 5 above)
Comment (alexcrichton) [reply src/cargo/ops/tree/graph.rs:68]: Nah seems reasonable to me, was just curious! It's documented now so seems fine by me either way.

File: src/doc/man/cargo-tree.adoc (hunk 14 above)
Comment (alexcrichton) [reply src/doc/man/cargo-tree.adoc]: Perhaps `--edges` may make the wording a bit nicer? That's sort of what it actually is (edges in a graph), but `--edges normal` doesn't read quite as nicely as `--dep-kinds normal`

File: src/doc/man/cargo-tree.adoc (hunk 15 above)
Comment (alexcrichton) [reply src/doc/man/cargo-tree.adoc:84]: I'm not overly familiar with `cargo tree` today, but given how nice I found this I'd expect `--graph-features` to be more common to justify commandeering `-f` from `--format`.

Thinking a bit though, I wonder if we could solve a bunch of birds with one stone? What if instead of `--graph-features` we did:

```
$ cargo tree --edges features
$ cargo tree -e features
```

(assuming we rename `--dep-kinds` to `--edges`). We could change it so the defaults of `--edges` is actually the empty string. We'd basically say that there are 4 kinds of edges (build/dev/normal/features) and everything but features is enabled by default. That way `--edges normal` would go away, but you'd instead do `--edges no-build,no-dev`.

Too radical?

**Comment by alexcrichton** [comment]
@ehuss oh my age is showing again, sorry `--workspace` is perfect and I should have been using that originally.

Although now testing it out, I'm not sure it's working? Or at least I'm getting [somewhat confusing output](https://gist.github.com/alexcrichton/848755cadb6e19858fcbc6da5f51c7b1) after executing:

```
$ cargo tree -p regex-syntax --graph-features -i --workspace
```

In that output the `regex-syntax` crate doesn't actually show up?

File: src/doc/man/cargo-tree.adoc (hunk 12 above)
Comment (alexcrichton) [reply src/doc/man/cargo-tree.adoc:48]: Oh right I forgot about `--duplicates`. Given the other consolidations that have been happening, I agree let's leave this as `--no-dedupe`

I like the idea though of changing `(*)` to something else. Do you think it'd be useful to take that a bit further and do `(deduped from line 13)` so if you pipe the output to a text editor you can jump around? Or something that points to where the deduplication happened at least.

**Comment by ehuss** [comment]
> `$ cargo tree -p regex-syntax --graph-features -i --workspace`

`-p` is ignored with `--workspace`.  We should probably have a warning about that.

I think in this case, you just want to run that same command without `--workspace`, though that will only give you the answer for one workspace member at a time (in wasmtime, it looks like `regex-syntax` is only used by the root package, so it doesn't matter there).

`-p` for a non-member is a bit of an oddball.  It can only select a non-member package inside the tree of the "current" member (the one in the current directory).  This is true of all other commands (`cargo build`, etc.).  That's likely not easy to change (I can explain more if you want).

**Comment by alexcrichton** [comment]
Ah right yeah, that makes sense. Do you know why, though, `cargo tree -i --workspace` only shows path dependencies? That seems like it may be a bit of a bug?

For my initial question `cargo tree --graph-features --workspace` plus "search the output" is what I wanted.

**Comment by ehuss** [comment]
> Do you know why, though, `cargo tree -i --workspace` only shows path dependencies?

It is because registry packages cannot depend on workspace members (ignoring `[patch]`).  For example, run `cargo tree --workspace`, and you can see the only things that depend directly on `witx` are `wig`, `wiggle`, `wiggle-generate`, and `wiggle-macro`.  So when you run `cargo tree --workspace -i`, only those 4 are listed as inverse dependencies.

**Comment by alexcrichton** [comment]
Hm ok so digging in a bit, I feel like the behavior with roots and `--invert` is a bit confusing. I don't think the `--workspace` flag is needed here, but `cargo tree` and `cargo tree -i`, for example, don't produce the same graph. With `cargo tree -i` you only get one package whereas with `cargo tree` you get the whole tree.

This makes sense with respect to the implementation, but it feels surprising to me and a bit counter-intuitive. Does it makes sense to invert the roots as well when we invert the graph? (so not just the edges). It does mean that `cargo tree -i -p some-cratesio-crate` doesn't make much sense.

Maybe what I'm looking for is root selection orthogonal to `-p`/`--workspace`/etc package selection. What I really want is `cargo tree --workspace -i --root regex-syntax` (terrible flag name). Or maybe `-i` doesn't make sense unless you explicitly specify some root you're interested in? 

Hm, do you agree that this is counter-intuitive in some situations? I don't really think this is a showstopper at all, just some thoughts I have from using it.

**Comment by ehuss** [comment]
I agree that it is confusing at first.  The first time I tried `cargo tree -i` and saw only one item, that was surprising.  But I don't know what else it could show that would make sense.  The whole point of `-i` is to show the edges flowing into a given node, and by default it chooses the current member which normally doesn't have any edges flowing into it.

I tried to emphasize in the documentation that `-i` usually needs `-p` to be useful.

Contemplating it more… We could change `-i` to take a package name.  Since it is only really useful with a named package, maybe it should emphasize that?  So maybe something like this:

* `cargo tree -i syn`:  Show what depends on `syn` for the current member.
* `cargo tree -i syn --workspace`:  Show what depends on `syn` across the entire workspace.
* `cargo tree -i syn -p cranelift`: Show what depends on `syn` inside the `cranelift` member.

And then bare `cargo tree -i` could just print a long help message explaining how to use it (since it is generally useless right now without arguments).  I could also probably make `cargo tree -p syn -i` work for backwards-compatibility.

I could also maybe add a hint to the error if you do `cargo tree -i foo` when `foo` isn't in the current member's tree, but exists elsewhere.

WDYT?

**Comment by alexcrichton** [comment]
I really like that idea, @sfackler in your time with `cargo-tree`, do you think that's a feasible change to make? (making `--invert` take an argument as to what you're interested in)

**Comment by sfackler** [comment]
So Gradle has a similar operation to what you do with `--invert`, and calls it `dependencyInsights` (in comparison to the normal `dependencies` task that acts like standard `cargo tree`. I like the idea of tailoring `--invert` to specifically taking a package, but it might be best to rename it to something else that's more descriptive.

File: src/doc/man/cargo-tree.adoc (hunk 12 above)
Comment (ehuss) [reply src/doc/man/cargo-tree.adoc:74]: Sounds interesting.  I'd like to put this on the future enhancements pile, though.

**Comment by ehuss** [comment]
I pushed an update that changes around some flags. I renamed `--dep-kinds` to `--edges` (or `-e`) and dropped `--graph-features`. It seems nicer, but I may have been staring at it for too long to have any perspective.

I also changed `-i` to take a package name. I think "invert" is still a good description of what the flag does.  I couldn't think of anything short to replace it.

**Comment by alexcrichton** [comment]
@rfcbot fcp merge

I'm game!

**Comment by alexcrichton** [comment]
@bors: r+

Ok I'm gonna go ahead and get the ball rolling on this, and we can always reevaluate if a blocking concern comes up!

**Comment by joshtriplett** [comment]
This is a minor nit, but I've never seen "dedupe" (singular) as an abbreviation for "deduplicate". I've always seen that written as `dedup` (which has the advantage of being a prefix of "deduplicate"). Would it be possible to rename the option before this is stabilized? (We could also have an alias.)

**Comment by ehuss** [comment]
I think an alias would be fine, though I think "dupe" is a well-established shorthand for duplicate. I also see more hits on google for "dedupe" than "dedup". Either is ok with me, though.

**Comment by rivy** [comment]
This hitting nightly just broke the `uutils/coreutils` CI build (see <https://github.com/rivy/rs.coreutils/runs/587889104?check_suite_focus=true>) because of the lack of support for the prior flags (specifically, hard faults for using `--all` or `-V`). I'm in the process of moving the CI to use beta for the time being (rust v1.43 is required). But, this took a while to track down and it's a hard, unexpected break for users of `cargo-tree`.

Maybe it would be better to just deprecate the old options with warnings until this hits at least beta or stable? Or, at least with all the warnings/errors, add some pointer to this PR so that users can more easily figure out the problem and possible solutions, specifically using `cargo-tree tree ...` as a workaround. That's an odd construction not easily otherwise discoverable without some pointer.

**Comment by ehuss** [comment]
@rivy Sure, I have opened #8115 to support those two flags.

**Comment by rivy** [comment]
@ehuss , thank you for the response and the rather quick PR solution!
